core/services/gateway/handlers/functions/subscriptions.go

```diff
@@ -0,0 +1,200 @@
+			return errors.New("OnchainSubscriptionsConfig.UpdateFrequencySec must be greater than 0")
+		}
+		if s.config.QueryTimeoutSec == 0 {
+			return errors.New("OnchainSubscriptionsConfig.UpdateTimeoutSec must be greater than 0")
+		}
+		if s.config.QueryRangeSize == 0 {
+			return errors.New("OnchainSubscriptionsConfig.QueryRangeSize must be greater than 0")
+		}
+
+		s.closeWait.Add(1)
+		go s.queryLoop()
+
+		return nil
+	})
+}
+
+func (s *onchainSubscriptions) Close() error {
+	return s.StopOnce("OnchainSubscriptions", func() (err error) {
+		s.lggr.Info("closing onchain subscriptions")
+		close(s.stopCh)
+		s.closeWait.Wait()
+		return nil
+	})
+}
+
+func (s *onchainSubscriptions) GetSubscription(address common.Address) *functions_router.IFunctionsSubscriptionsSubscription {
+	s.rwMutex.RLock()
+	defer s.rwMutex.RUnlock()
+	subscription, ok := s.subscriptions[address]
+	if !ok {
+		return nil
+	}
+	return subscription
+}
+
+func (s *onchainSubscriptions) queryLoop() {
+	defer s.closeWait.Done()
+
+	ticker := time.NewTicker(time.Duration(s.config.QueryFrequencySec) * time.Second)
+	defer ticker.Stop()
+
+	var start uint64 = 1
+
+	queryFunc := func() {
+		ctx, cancel := utils.ContextFromChanWithTimeout(s.stopCh, time.Duration(s.config.QueryTimeoutSec)*time.Second)
+		defer cancel()
+
+		latestBlockHeight, err := s.client.LatestBlockHeight(ctx)
+		if err != nil || latestBlockHeight == nil {
+			s.lggr.Errorw("Error calling LatestBlockHeight", "err", err, "latestBlockHeight", latestBlockHeight.Int64())
+			return
+		}
+
+		blockNumber := big.NewInt(0).Sub(latestBlockHeight, s.blockConfirmations)
+
+		count, err := s.getSubscriptionsCount(ctx, blockNumber)
+		if err != nil {
+			s.lggr.Errorw("Error getting subscriptions count", "err", err)
+			return
+		}
+		if count == 0 {
+			s.lggr.Info("Router has no subscriptions yet")
+			return
+		}
+
+		if start > count {
+			start = 1
+		}
+
+		end := start + uint64(s.config.QueryRangeSize)
+		if end > count {
+			end = count
+		}
+		if err := s.querySubscriptionsRange(ctx, blockNumber, start, end); err != nil {
+			s.lggr.Errorw("Error querying subscriptions", "err", err, "start", start, "end", end)
+			return
+		}
+
+		start = end + 1
+	}
+
+	for {
+		select {
+		case <-s.stopCh:
+			return
+		case <-ticker.C:
+			queryFunc()
+		}
+	}
+}
+
+func (s *onchainSubscriptions) querySubscriptionsRange(ctx context.Context, blockNumber *big.Int, start, end uint64) error {
+	subscriptions, err := s.router.GetSubscriptionsInRange(&bind.CallOpts{
+		Pending:     false,
+		BlockNumber: blockNumber,
+		Context:     ctx,
+	}, start, end)
+	if err != nil {
+		return errors.Wrap(err, "unexpected error during functions_router.GetSubscriptionsInRange")
+	}
+
+	s.rwMutex.Lock()
+	defer s.rwMutex.Unlock()
+	for _, subscription := range subscriptions {
+		if subscription.Owner == utils.ZeroAddress {
+			continue
+		}
+		subscription := subscription
+		s.subscriptions[subscription.Owner] = &subscription
```

Comment:
Maybe we should check the balance and only save the subscription with the highest balance here? Although maybe we should also re-work this as we are currently holding the rwMutext here.

Reply:
Let's save all but have an API to retrieve highest balance. In the future we are also going to look at flags stored within subscriptions.